Add VDEL rules to Conway

aniketd opened the pull request:
# Description

- [ ] ~~Rename all `CommitteeCert`s to `GovCert`s~~
- [ ] ~~Rename `TALLY` rule to `GOV` cc. @lehins @Soupstraw @WhatisRT (perhaps in a separate issue and PR)~~
- [x] The renames may be done in another follow-up PR.

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated
- [x] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [x] The version bounds in `.cabal` files for all affected packages are updated (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/input-output-hk/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

aniketd committed:
Add VDEL rules to Conway

Soupstraw commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/VDel.hs:
```diff
@@ -3,63 +3,110 @@
+  case c of
+    ConwayVDelRegDRep cred -> do
+      Set.notMember cred vsDReps ?! ConwayDRepAlreadyRegisteredVDEL cred
+      -- NOTE: @Soupstraw: something about DRep Deposits seems to be missing here.
```
We don't currently have a protocol parameter for DRep deposits, we should at some point add it and then check against that instead of using the poolDeposit parameter. I think it's a good idea to leave a TODO here and then revisit it once we've updated PParams

Soupstraw commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/VDel.hs:
```diff
@@ -3,63 +3,110 @@
+          { vsDReps = Set.delete cred vsDReps
+          }
+    ConwayVDelRegCommittee coldK hotK -> do
+      -- NOTE: @Soupstraw: from the spec: "Should we check if KeyHash actually belongs to the CC?"
```
Leave this as a TODO, we'll revisit this once we've figured it out in the spec.

Soupstraw commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/VDel.hs:
```diff
@@ -3,63 +3,110 @@
 data ConwayVDelPredFailure era
-  deriving (Show, Eq, Generic, NoThunks, NFData)
+  = ConwayDRepAlreadyRegisteredVDEL !(Credential 'Voting (EraCrypto era))
+  | ConwayCommitteeAlreadyRegisteredVDEL !(KeyHash 'CommitteeColdKey (EraCrypto era))
```
I'm not sure we want to add `VDEL` to the end of every failure. Maybe @lehins can comment on the naming practices.

aniketd replied:
I think this is the pattern in all the rules. It will soon be renamed to `GOVCERT` though, as per discussions on Slack :+1:

Soupstraw commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/VDel.hs:
```diff
@@ -3,63 +3,110 @@
+        vState
+          { vsDReps = Set.insert cred vsDReps
+          }
+    ConwayVDelUnRegDRep cred ->
```
Here we should check that the credential is in the DReps map. I think the spec has a mistake at this place, I'll fix that.

aniketd replied:
Got it!

Soupstraw commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/VDel.hs:
```diff
@@ -3,63 +3,110 @@
+          }
+    ConwayVDelRegCommittee coldK hotK -> do
+      -- NOTE: @Soupstraw: from the spec: "Should we check if KeyHash actually belongs to the CC?"
+      Map.notMember coldK vsCCHotKeys ?! ConwayCommitteeAlreadyRegisteredVDEL coldK
```
We might want to allow re-registering an already exisiting key, the spec is not very clear about this. I'll look into it.

aniketd replied:
Cool! Let me know if I need to take this out.

Soupstraw replied:
yeah, it seems like we'll want to allow re-registering without having to unregister first, so this check can be removed.

Soupstraw commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/VDel.hs:
```diff
@@ -3,63 +3,110 @@
+        vState
+          { vsCCHotKeys = Map.insert coldK hotK vsCCHotKeys
+          }
+    ConwayVDelUnRegCommittee coldK ->
```
Here we probably want to check that the cold key is registered.

aniketd replied:
Got you!

Soupstraw commented on libs/cardano-ledger-core/src/Cardano/Ledger/CertState.hs:
```diff
@@ -257,8 +257,8 @@ data VState era = VState
-          (KeyHash 'Voting (EraCrypto era))
-          (KeyHash 'Voting (EraCrypto era))
+          (KeyHash 'CommitteeColdKey (EraCrypto era))
+          (KeyHash 'CommitteeHotKey (EraCrypto era))
```
nice!

aniketd committed:
Add patterns and CBOR instances

aniketd committed:
Update changelog

Soupstraw commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Cert.hs:
```diff
@@ -125,8 +126,8 @@ certTransition = do
       pure $ cState {certPState = newPState}
-    ConwayTxCertCommittee committeeCert -> do
-      newVState <- trans @(EraRule "VDEL" era) $ TRC (VDelEnv, certVState, committeeCert)
+    ConwayTxCertCommittee vDelCert -> do
```
I think we can also rename the `ConwayTxCertCommittee` constructor.

aniketd replied:
Can I do this along with the other the renames together in a follow up PR?

WhatisRT commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/VDel.hs:
```diff
@@ -12,44 +13,57 @@
-  deriving (Show, Eq, Generic, NoThunks, NFData)
+  = ConwayDRepAlreadyRegisteredVDEL !(Credential 'Voting (EraCrypto era))
+  | ConwayDRepNotRegisteredVDEL !(Credential 'Voting (EraCrypto era))
+  | ConwayCommitteeAlreadyRegisteredVDEL !(KeyHash 'CommitteeColdKey (EraCrypto era))
```
This isn't necessary. The point of the hot/cold key scheme is to be able to change the hot key, so being already registered cannot be a failure.

The error message right below also isn't right. The only way registering can fail is if that CC member has resigned. So one error named something like `ConwayCommitteeHasResignedVDEL` would be appropriate.

WhatisRT commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/VDel.hs:
```diff
@@ -58,9 +72,44 @@ instance
+          { vsDReps = Set.delete cred vsDReps
+          }
+    ConwayAuthCommitteeHotKey coldK hotK -> do
+      -- TODO: implement after the spec is to be updated w.r.t. checking if the coldK belongs to CC.
```
The two rules here should have identical logic. They fail iff the CC member is already resigned. So `vsCCHotKeys` needs to accommodate that somebody has resigned. In the spec this is done with inserting a `nothing`, so `Map.delete` below needs to be changed.

Also, this TODO comment can be removed - we won't check is the key is in the CC. Leaving that check out actually speeds up changing the CC by an epoch.

WhatisRT reviewed (commented):
I don't think I can say anything useful beyond what's in `VDel.hs`, so I'm not sure how useful my review is.

The details of this rule have changed a bit (see [here](https://github.com/input-output-hk/formal-ledger-specifications/pull/123/files#diff-2052333992aa192efb62901ebc339a8c4bcf6cb2001c9015d4ecdb4727210995)) but I don't know if you want to incorporate all of that here before that PR has even gotten a review.

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/TxCert.hs:
```diff
@@ -291,6 +329,7 @@ committeeKeyHashWitness :: ConwayCommitteeCert c -> KeyHash 'Witness c
 committeeKeyHashWitness = \case
   ConwayAuthCommitteeHotKey coldKeyHash _ -> asWitness coldKeyHash
   ConwayResignCommitteeColdKey coldKeyHash -> asWitness coldKeyHash
+  _ -> undefined
```
Due to changes to the cert type, this function now needs to return a Maybe
```suggestion
  _ -> Nothing
```

aniketd replied:
Got you!

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/TxCert.hs:
```diff
@@ -279,7 +315,9 @@ instance NFData (ConwayDelegCert c)
 
 data ConwayCommitteeCert c
-  = ConwayAuthCommitteeHotKey !(KeyHash 'CommitteeColdKey c) !(KeyHash 'CommitteeHotKey c)
+  = ConwayDRepReg !(Credential 'Voting c)
```
I think it would be good to add deposits to the reg/unreg certificates in this PR. This would reduce number of times we have to touch CBOR implementation and cddl spec

aniketd replied:
@WhatisRT should we also store the deposits somewhere in the `VState`?

lehins replied:
Yes, we'll need to store deposits in the VState, but we can take care of handling deposits with a later PR. I am only suggesting adding the `Coin` fields to `ConwayDRepReg` and `ConwayDRepUnReg`, just so at least the types are ready.

aniketd replied:
Yes, thanks @lehins ! I have added them but not using them to make any checks etc. yet :+1:

lehins replied:
Also for consistency sake they should be named:
`ConwayRegDRep` and `ConwayUnRegDRep`

aniketd replied:
Got it!

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/CertState.hs:
```diff
@@ -257,8 +257,8 @@ data VState era = VState
   { vsDReps :: !(Set (Credential 'Voting (EraCrypto era)))
   , vsCCHotKeys ::
```
This will be good for consistency
```suggestion
  , vsCommitteeHotKeys ::
```

lehins reviewed (changes requested):
Looks good. It's definitely ok, to merge this without it being 100% finished. All the discussed renames can be done in a follow up PR.

aniketd committed:
Address review comments.

- Change VState ColdHot delegation map range to Maybe
- Add deposit fields to Conway(Un)RegDRep certs

aniketd committed:
Rename vsCCHotKeys to vsCommitteeHotKeys

aniketd committed:
Fix Conway DRep Cert names

aniketd commented:
@WhatisRT 

> The details of this rule have changed a bit (see [here](https://github.com/input-output-hk/formal-ledger-specifications/pull/123/files#diff-2052333992aa192efb62901ebc339a8c4bcf6cb2001c9015d4ecdb4727210995)) but I don't know if you want to incorporate all of that here before that PR has even gotten a review.

I have not incorporated those changes here :+1: . But, the changes you and @lehins  suggested have been done here now, and this is ready for your final review. :raised_hands:

WhatisRT commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/VDel.hs:
```diff
@@ -58,9 +75,34 @@ instance
+      (isNothing <$> Map.lookup coldK vsCommitteeHotKeys) /= Just True ?! ConwayCommitteeHasResignedVDEL coldK
+      pure $ vState {vsCommitteeHotKeys = Map.insert coldK (Just hotK) vsCommitteeHotKeys}
+    ConwayResignCommitteeColdKey coldK -> do
+      Map.member coldK vsCommitteeHotKeys ?! ConwayCommitteeNotRegisteredVDEL coldK
```
This line and the corresponding error message should be removed. There is no reason why someone shouldn't be allowed to immediately resign.

aniketd replied:
Oh okay, I will remove it now. This would allow a person who hasn't ever authorized a hot key, to simply resign. Thanks @WhatisRT :raised_hands:

lehins replied:
@WhatisRT What if committee member already resigned, double resign is likely a mistake, so checking for it might be good, right?
```suggestion
       validateCommitteeColdKeyHasNotResigneed coldK vsCommitteeHotKeys
```

where

```haskell
validateCommitteeColdKeyHasNotResigneed coldK vsCommitteeHotKeys =
 (isNothing <$> Map.lookup coldK vsCommitteeHotKeys) /= Just True ?! ConwayCommitteeHasResignedVDEL coldK
```

WhatisRT replied:
Yeah, it doesn't really matter since it doesn't change the state. But the logic in the spec is the same, if you've resigned you can't publish any certs for that cold key anymore. So that check makes sense to keep.

aniketd replied:
Updated! :raised_hands:

aniketd committed:
Remove unnecessary membership check for ResignCommitteeKey

WhatisRT reviewed (approved):
LGTM

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/TxCert.hs:
```diff
@@ -347,6 +387,14 @@ conwayTxCertDelegDecoder = \case
+  17 -> do
+    cred <- decCBOR
+    deposit <- decCBOR
+    pure (3, UnRegDRepTxCert cred deposit)
```
We also need to adjust cddl spec for conway certificates. It's ok to add it in a subsequent PR too.

aniketd replied:
Awesome catch! Hope I have done the right thing there now :crossed_fingers:

lehins commented on eras/conway/impl/src/Cardano/Ledger/Conway/Rules/VDel.hs:
```diff
@@ -12,44 +13,60 @@
+  | ConwayDRepNotRegisteredVDEL !(Credential 'Voting (EraCrypto era))
+  | ConwayDRepIncorrectDepositVDEL !Coin
+  | ConwayCommitteeHasResignedVDEL !(KeyHash 'CommitteeColdKey (EraCrypto era))
+  | ConwayCommitteeNotRegisteredVDEL !(KeyHash 'CommitteeColdKey (EraCrypto era))
```
You might wanna rebase on this PR and add Enc/DecCBOR instances for this type
https://github.com/input-output-hk/cardano-ledger/pull/3477

Soon above PR will get merged and you'll have to do it anyways, so it is just a heads up

aniketd replied:
Oh yes, I saw @teodanciu PR just now. Got it! I will rebase on top of it and update here. :raised_hands: Thanks Alexey!

lehins reviewed (approved):
Very nice!

aniketd committed:
Rebase on master; add CBOR instance; update CDDL

lehins reviewed (approved):
Beautiful!
